core/src/main/java/io/confluent/rest/Application.java

```diff
@@ -256,52 +256,47 @@ protected void doStop() throws Exception {
-    defaultHolder.setInitParameter("dirAllowed", "false");
+    configureResourceExtensions(resourceConfig);
+    configureBaseApplication(resourceConfig, combinedMetricsTags);
 
-    ResourceCollection staticResources = getStaticResources();
-    if (staticResources != null) {
-      context.setBaseResource(staticResources);
+    // Support mono context setup
+    if (appCtx.size() == 0) {
+      appCtx.add(new DefaultContext(resourceConfig));
     }
 
-    configureSecurityHandler(context);
+    HandlerCollection handlers = new HandlerCollection();
+    for (ApplicationContext ctx : appCtx) {
 
-    if (isCorsEnabled()) {
-      String allowedOrigins = config.getString(RestConfig.ACCESS_CONTROL_ALLOW_ORIGIN_CONFIG);
-      FilterHolder filterHolder = new FilterHolder(CrossOriginFilter.class);
-      filterHolder.setName("cross-origin");
-      filterHolder.setInitParameter(
-          CrossOriginFilter.ALLOWED_ORIGINS_PARAM, allowedOrigins
+      // Isolate contexts by making a defensive copy of the application config
+      ResourceConfig ctxConfig = ctx.ctxConfig;
 
-      );
-      String allowedMethods = config.getString(RestConfig.ACCESS_CONTROL_ALLOW_METHODS);
-      String allowedHeaders = config.getString(RestConfig.ACCESS_CONTROL_ALLOW_HEADERS);
-      if (allowedMethods != null && !allowedMethods.trim().isEmpty()) {
-        filterHolder.setInitParameter(CrossOriginFilter.ALLOWED_METHODS_PARAM, allowedMethods);
-      }
-      if (allowedHeaders != null && !allowedHeaders.trim().isEmpty()) {
-        filterHolder.setInitParameter(CrossOriginFilter.ALLOWED_HEADERS_PARAM, allowedHeaders);
-      }
-      // handle preflight cors requests at the filter level, do not forward down the filter chain
-      filterHolder.setInitParameter(CrossOriginFilter.CHAIN_PREFLIGHT_PARAM, "false");
-      context.addFilter(filterHolder, "/*", EnumSet.of(DispatcherType.REQUEST));
-    }
-    configurePreResourceHandling(context);
-    context.addFilter(servletHolder, "/*", null);
-    configurePostResourceHandling(context);
-    context.addServlet(defaultHolder, "/*");
 
-    applyCustomConfiguration(context, REST_SERVLET_INITIALIZERS_CLASSES_CONFIG);
+      ctx.setBaseResource(ctx.getStaticResources());
+
+      ctx.setupResources(ctxConfig, ctx.getConfiguration());
+      ctx.configureSecurityHandler();
+      ctx.enableCors();
+
+      ctx.configurePreResourceHandling();
+      ServletContainer servletContainer = new ServletContainer(ctxConfig);
+      FilterHolder servletHolder = new FilterHolder(servletContainer);
+      ctx.addFilter(servletHolder, "/*", null);
+
+      ctx.configurePostResourceHandling();
+      ServletHolder defaultHolder = new ServletHolder("default", DefaultServlet.class);
+      defaultHolder.setInitParameter("dirAllowed", "false");
+      ctx.addServlet(defaultHolder, "/*");
+
+      applyCustomConfiguration(ctx, REST_SERVLET_INITIALIZERS_CLASSES_CONFIG);
+      handlers.addHandler(ctx);
```

Comment:
Could all this block be in ApplicationContext, and do handlers.addHandler(ctx.configureHandler()) or something similar?